### PR TITLE
Replace BIO_printf with ASN1_STRING_print in GENERAL_NAME_print

### LINKED ISSUE #8997: GENERAL_NAME_print mishandled embedded nuls in IA5 string values
See [https://github.com/openssl/openssl/blob/master/crypto/x509v3/v3_alt.c](url)

```
    case GEN_EMAIL:
        BIO_printf(out, "email:%s", gen->d.ia5->data);
        break;

    case GEN_DNS:
        BIO_printf(out, "DNS:%s", gen->d.ia5->data);
        break;

    case GEN_URI:
        BIO_printf(out, "URI:%s", gen->d.ia5->data);
        break;
```

Each of these instances should be using ASN1_STRING_print to print the value rather than printing the raw ia5 string data. If there is an embedded end-of-string (\x00) character the user will be unaware.

For code using GENERAL_NAME_print to handle the various formats and then post-processing for checking logic, the wrong handling will occur.

For users working with the command line applications, incorrect output will occur.

If we have embedded end-of-string values in the subject or issuer name, those are displayed correctly. For general ASN1_STRING handling we fold special characters to a period (.) which is another way of indicating there is more data. 

Reproduce the issue with a certificate created with a SAN with DNS and an embedded \x00 within the DNS name (yes this is an invalid usage - however we shouldn't silently truncate the output which may lead to unexpected handling).

### PR DESCRIPTION
**Pull request by acheevbhagat**
Fixes #8997 
CLA: trivial

### CODE REVIEW

**Comment by paulidale** [comment]
These changes are not trivial.  The [CLA process](https://www.openssl.org/policies/cla.html) will need to be followed.

File: crypto/x509v3/v3_alt.c [hunk 1]
```diff
@@ -157,15 +159,36 @@ int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
...
     case GEN_EMAIL:
-        BIO_printf(out, "email:%s", gen->d.ia5->data);
-        break;
+	formatted_output = ASN1_STRING_new();
+	formatted_length = sizeof(gen->d.ia5->data);
+	data = calloc(7 + formatted_length, sizeof(char)); 
+	snprintf(data, 7 + formatted_length, "email:%s", gen->d.ia5->data);
+	ASN1_STRING_set(formatted_output, data, 7 + formatted_length);
+	ASN1_STRING_print(out, formatted_output);
+	ASN1_STRING_free(formatted_output);
```
Comment (t-j-h) [inline crypto/x509v3/v3_alt.c]: No - the approach should be a BIO_printf of the label (i.e. "email") followed by a ASN1_STRING_print.
There is no need to create a *copy* of the data and prepend a label.

GENERAL_NAME_get0_value can be used to return the ia5 value - or you can directly reference it given the current code does that. It is an ASN1_IA5STRING which is an ASN1_STRING ...

e.g. something like the following 

```
BIO_printf(out,"email:");
ASN1_STRING_print(out,gen->d.ia5);

```

File: crypto/x509v3/v3_alt.c [hunk 2]
```diff
@@ -141,7 +141,9 @@ STACK_OF(CONF_VALUE) *i2v_GENERAL_NAME(X509V3_EXT_METHOD *method,
 int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
 {
     unsigned char *p;
-    int i;
+    char *data;
+    int i, formatted_length;
+    ASN1_STRING *formatted_output;
     switch (gen->type) {
```
Comment (t-j-h) [inline crypto/x509v3/v3_alt.c]: Blank line required between vars and code

Comment (acheevbhagat) [reply crypto/x509v3/v3_alt.c]: I'll add that in

File: crypto/x509v3/v3_alt.c (hunk 1 above)
Comment (acheevbhagat) [reply crypto/x509v3/v3_alt.c]: I see, I had a misunderstanding of how it worked, I thought the out BIO would have to take it from one print statement. I'll change it.

**Comment by acheevbhagat** [comment]
I'll email in my ICLA tonight.

**Comment by acheevbhagat** [comment]
I've put in the recommended changes and submitted the ICLA

**Comment by mattcaswell** [comment]
@acheevbhagat Why did you close this, it has not yet been merged?

**Comment by acheevbhagat** [comment]
Because I'd messed up the whole log pretty bad trying to squash the commits. I'm not really that great at git. I deleted and recreated my fork, and ported over my changes to the new fork and I'm gonna put in a new pull request, is that alright?

**Comment by acheevbhagat** [comment]
I've submitted the new PR #9101

**Comment by mattcaswell** [comment]
> Because I'd messed up the whole log pretty bad trying to squash the commits. I'm not really that great at git. I deleted and recreated my fork, and ported over my changes to the new fork and I'm gonna put in a new pull request, is that alright?

We'll have to move the approvals from here over to the new PR

**Pull request closed without merging**